internal/security/spiffetokenprovider/init.go

```diff
@@ -0,0 +1,344 @@
+	})
+
+	//
+	// The SPIFFE token provider will take three parameters:
+	//
+	// An X.509 SVID used in mutual-auth TLS for the token provider and the service to cross-authenticate.
+	//
+	// The requested service key. If blank, the service key will default to the service name encoded in the SVID.
+	// If the service name follows the pattern device-(name), then the service key must follow the format
+	// device-(name) or device-name-*. If the service name is app-service-configurable,
+	// then the service key must follow the format app-*. (This is an accommodation for the Unix workload
+	// attester not being able to distingish workloads that are launched using the same executable binary.
+	// Custom app services that support multiple instances won't be supported unless they name the executable
+	// the same as the standard app service binary or modify this logic.)
+	//
+	// A list of "known secret" identifiers that will allow new services to request database passwords or other
+	// "known secrets" to be seeded into their service's partition in the secret store.
+	//
+
+	// Handle gettoken endpoint
+	http.HandleFunc("/api/v2/gettoken", func(w http.ResponseWriter, r *http.Request) {
+		lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+
+		// Have to read the connection to finish TLS verification
+		if err := r.ParseForm(); err != nil {
+			lc.Errorf("Could not parse form: %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		serviceKey := r.Form.Get("service_key")
+		// could get multiple secret names from the form posted. e.g.: redisdb,mongodb,foo
+		// for the known secrets we currently only supports redisdb
+		// Due to the service key from the form only sees one service for the current use case
+		knownSecretNames := r.Form["known_secret_names"]
+		rawTokenOnly, err := strconv.ParseBool(r.Form.Get("raw_token"))
+		if err != nil {
+			lc.Errorf("Could not parse bool '%s': %v", r.Form.Get("raw_token"), err)
+		}
+		peerSVID := ""
+
+	iterateCertificates:
+		for _, cert := range r.TLS.PeerCertificates {
+			for _, uri := range cert.URIs {
+				// First certificate in verified chain has URI list,
+				// one of which is the SVID containing the EdgeX service key
+				if strings.HasPrefix(uri.String(), "spiffe://") {
+					peerSVID = uri.String()
+					break iterateCertificates
+				}
+			}
+		}
+
+		fmt.Println("PeerSVID:", peerSVID)
+		// TODO: verify the prefix with what we expect like spiffe://edgexfoundry.org/service/*
+		// regex: regexp.MustCompile(`(?m)^(?:spiffe://)(.)*/service/(.)*$`)
+		spiffeServiceKey := strings.Replace(peerSVID,
+			fmt.Sprintf("spiffe://%s/service/", configuration.Spiffe.TrustDomain.HostName), "", 1)
+
+		privilegedToken, err := b.getPrivilegedToken(dic)
+		if err != nil {
+			lc.Errorf("failed to load secret store token: %v", err)
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		vaultTokenResponse, err := makeToken(spiffeServiceKey, privilegedToken, secretStoreClient, lc)
```

Comment:
Before making the `makeToken` call, we need to apply the following logic from the ADR.  In the description below, "service name" is the service that was parsed from the SPIFFE ID using the regexp.  The "service key" is the form parameter that was passed by the client in the request.

---

The reqested service key. If blank, the service key will default to the service name encoded in the SVID. If the service name follows the pattern device-(name), then the service key must follow the format device-(name) or device-name-*. If the service name is app-service-configurable, then the service key must follow the format app-*. (This is an accomodation for the Unix workload attester not being able to distingish workloads that are launched using the same executable binary. Custom app services that support multiple instances won't be supported unless they name the executable the same as the standard app service binary or modify this logic.)